pkg/operator/ceph/controller/predicate.go

```diff
@@ -80,6 +94,10 @@ func WatchControllerPredicate() predicate.Funcs {
 				diff := cmp.Diff(objOld.Spec, objNew.Spec, resourceQtyComparer)
 				if diff != "" {
 					logger.Infof("CR has changed for %q. diff=%s", objNew.Name, diff)
+					// restart the rook operator so already created bucket claims can be reconciled
+					if err := restartOperator(); err != nil {
```

Comment:
Per offline discussion, we need to find a more scoped fix instead of restarting all the controllers.

Reply:
I'm going to break this down for the CephObjectStoreUser case first:

In the CephObjectStoreUser (COSU) controller, we need to enqueue **new** COSU requests when a CephObjectStore (COS) that is the parent changes its endpoint. The "Watching Arbitrary Resources" section of the kubebuilder book [here](https://book-v1.book.kubebuilder.io/beyond_basics/controller_watches.html) has more details on how to do that.

Essentially what we need to do is watch COS'es from the COSU controller and create a "map function" that takes a COS as an input and returns a list of COSU's that are affected by the input COS.

The map function is pretty easy. We merely need to list all COSU's that have the COS as a parent. [CRDs don't support arbitrary field selection](https://github.com/kubernetes/kubernetes/issues/53459), so we are going to just have to list all COSU resources and return a list only the NamespacedNames that refer to the given COS.

The predicate is the hardest part to design as far as logic goes. We need to make sure we don't miss any edge cases, and we would like to make sure not to use this code loop more often than necessary due to the expense of listing all COSU's.

We know we need to handle COS updates, but what about Create/Delete? Delete is unnecessary because rook blocks COS deletion until all COSU's are also deleted. Add is unnecessary in **both** cases where we see Create events:
1. In the event of a newly created COS, any COSU's that use it will block provisioning until the COS is ready. 
2. In the event of an operator restart where **all** resources get Create events, the operator is already going to reconcile all COSU's.

Therefore, the predicate options that handle COS'es should be designed like this:
- Return false from Create and Delete events
- Use the following process for Update events (both old and new are available)
  - IF new COS doesn't have a status with an endpoint, return false to ignore
  - IF old COS doesn't have a status with an endpoint, also return false to ignore. This is because this must be a newly-created COS that just finished reconciling, and any COSU's will already be waiting in a queue to reconcile.
  - IF the new COS endpoint is the same as the old COS endpoint, return false to ignore updates that don't affect COSU's
  - now return true to allow the map function to enqueue any COSU's that are children of this COS

Reply:
The pkg/operator/ceph/object/bucket reconciler for OBCs also uses controller-runtime. We should do the exact same thing for OBCs as we do for COSU's that I outlined above. The only change is the map function, actually.

Reply:
THanks for the great suggestion.